public/app/plugins/datasource/loki/queryUtils.ts

```diff
@@ -112,6 +112,12 @@ export function getNormalizedLokiQuery(query: LokiQuery): LokiQuery {
   return { ...rest, queryType: LokiQueryType.Range };
 }
 
+// This is a wrapper over getNormalizedLokiQuery to make returned query type non-nullable
+export function getLokiQueryType(query: LokiQuery): LokiQueryType {
+  const defaultQueryType = LokiQueryType.Range;
+  return getNormalizedLokiQuery(query).queryType ?? defaultQueryType;
+}
+
```

Comment:
i see, the fact that LokiQuery.queryType can be undefined is annoying. what do you think of turning this completely around like this?
```ts
// we are migrating from `.instant` and `.range` to `.queryType`
// this function returns a new query object that:
// - has `.queryType`
// - does not have `.instant`
// - does not have `.range`
export function getNormalizedLokiQuery(query: LokiQuery): LokiQuery {
  const queryType = getLokiQueryType(query);
  const { instant, range, ...rest } = query;
  return { ...rest, queryType };
}

// we are migrating from `.instant` and `.range` to `.queryType`
// this function returns the correct query type
export function getLokiQueryType(query: LokiQuery): LokiQueryType {
  //  if queryType field contains invalid data we behave as if the queryType is empty
  const { queryType } = query;
  const hasValidQueryType =
    queryType === LokiQueryType.Range || queryType === LokiQueryType.Instant || queryType === LokiQueryType.Stream;

  // if queryType exists, it is respected
  if (hasValidQueryType) {
    return queryType;
  }

  // if no queryType, and instant===true, it's instant
  if (query.instant === true) {
    return LokiQueryType.Instant;
  }

  // otherwise it is range
  return LokiQueryType.Range;
}
```

Reply:
yeah make sense

Reply:
https://github.com/grafana/grafana/pull/70237/commits/4153a5b1083348d533f170ffc61716308d454a02